### ISSUE TITLE
Support for Materialized Views

### ISSUE THREAD

**Issue by proddata** [crate/crate#10661]
**Use case**:
As a user I want to be able to use some sort of materialized views or continuous aggregates to speed up query times on larger aggregated queries. Updates to those queries can be triggered manually or by an external cron job.

**Feature description**:
Implementation optimally would follow [PostgreSQL implementation](https://www.postgresql.org/docs/current/rules-materializedviews.html)

Creating materialized view:
```sql
CREATE MATERIALIZED VIEW materialized_view AS
   SELECT device, max(value) FROM my_table;
```

Updating materialized view:
```sql
REFRESH MATERIALIZED VIEW materialized_view;
```

see also: https://github.com/crate/crate/issues/8806

**Comment by proddata** [comment]
The described workaround (`CREATE TABLE / INSERT INTO`) has some flaws:
- primary keys on the materialisation table are required for `INSERT INTO`
- during updates, there are inconsistencies in the result

A workaround for "simple" materialization would be:

```sql
/* CREATE MATERIALIZED VIEW my_table_rollup AS
SELECT
    device_id,
    date_trunc('week',ts) week,
    SUM(val) avg_val
FROM my_table
GROUP BY 1, 2;
*/
CREATE TABLE my_table_rollup AS
SELECT
    device_id,
    date_trunc('week',ts) week,
    SUM(val) avg_val
FROM my_table
GROUP BY 1, 2;
```
and then update via

```sql
/* REFRESH MATERIALIZED VIEW my_table_rollup; */
CREATE TABLE temp_my_table_rollup AS
SELECT
    device_id,
    date_trunc('week',ts) week,
    AVG(val) avg_val
FROM my_table
GROUP BY 1, 2;
ALTER CLUSTER SWAP TABLE temp_my_table_rollup TO my_table_rollup WITH ("drop_source" = true);
```

While this might not look as too much effort in the first place, managing especially the update part is rather cumbersome, especially when one has to deal with updating 10,20,30, ... "views" with some sort of cron job.

Replacing view definitions, would not only need the change in the "view" (table), but also the mechanism that updates it accordingly. Especially in organisation, where this might not be handled by the same person, this becomes tedious.

**Comment by proddata** [comment]
a better approach might be to use a partitioned table with a view on top of it, which gets updated only after the insert into a new partition was successful.

for a view like
```sql
CREATE VIEW my_table_rollup AS
SELECT
    device_id,
    date_trunc('week',ts) week,
    SUM(val) avg_val
FROM my_table
GROUP BY 1, 2;
```

one could use the following approach:

```sql
CREATE TABLE my_table_rollup_materialization (
   update TIMESTAMP,
   device_id TEXT,
   week TIMESTAMP,
   avg_val DOUBLE
) PARTITIONED BY(update);

-- create timestamp e.g. ts_now
INSERT INTO my_table_rollup_materialization SELECT <ts_now>, device_id,  week, avg_val FROM my_table;
REFRESH TABLE my_table_rollup_materialization;

-- if successful
CREATE OR REPLACE VIEW my_table_rollup_materialized AS 
SELECT
    device_id,
    week,
    avg_val
FROM my_table_rollup_materialization
WHERE update = <ts_now>;
DELETE FROM my_table_rollup_materialized WHERE update < <ts_now>
```

of course it would be much nicer to have syntactic sugar for this within CrateDB

**Comment by ckurze** [comment]
Another option would be the implementation of something like a MERGE statement, i.e. recalculate the new data in the source table and merge it into a target table:

https://docs.oracle.com/en/database/oracle/oracle-database/12.2/sqlrf/MERGE.html#GUID-5692CCB7-24D9-4C0E-81A7-A22436DC968F 

MERGE INTO bonuses D
   USING (SELECT employee_id, salary, department_id FROM hr.employees
   WHERE department_id = 80) S
   ON (D.employee_id = S.employee_id)
   WHEN MATCHED THEN UPDATE SET D.bonus = D.bonus + S.salary*.01
     DELETE WHERE (S.salary > 8000)
   WHEN NOT MATCHED THEN INSERT (D.employee_id, D.bonus)
     VALUES (S.employee_id, S.salary*.01)
     WHERE (S.salary <= 8000);

**Comment by proddata** [comment]
> Another option would be the implementation of something like a MERGE statement, i.e. recalculate the new data in the source table and merge it into a target table:

Is not really an alternative to materialized views, but rather to `INSERT INTO ... ON CONFLICT SET ...` which already cover a lot of the `MERGE` use cases.

**Comment by proddata** [comment]
**Draft proposal following PostgreSQL beahviour**
⚠️ Feel free to edit
⚠️ No decision to implement

The propose approach largely follows the [PostgreSQL implementation of materialized views](https://www.postgresql.org/docs/current/sql-creatematerializedview.html). 

> `CREATE MATERIALIZED VIEW` is similar to` CREATE TABLE AS`, except that it also remembers the query used to initialize the view, so that it can be refreshed later upon demand. A materialized view has many of the same properties as a table, but there is no support for temporary materialized views. Criteria

- A User can define a materialized view using a CREATE MATERIALIZED VIEW statement.
```sql
CREATE MATERIALIZED VIEW my_mat_view AS SELECT * FROM my_table;
```
- A User can optionally specify column definitions for the materialised views.
- If they are not specified, the behaviour is the same as CREATE TABLE AS
- The `SELECT` query is executed and used to populate the view right after the view is created.
- The statement requires DDL  privileges.
- A User can update/refresh the data in a materialized view using REFRESH MATERIALIZED VIEW.
```sql
REFRESH MATERIALIZED VIEW my_mat_view;
```
- The refresh statement can either succeed or fail and return an appropriate value.
- If the refresh fails, the latest full refresh is kept.
- If the refresh succeeds, all other data is cleaned up.
- There is the option to specify a number of replicas for a materialized view with WITH (number_of_replicas = '0-1')  to specify how often it is replicated.
- There is the option to ignore the statement if the view already exits: IF NOT EXISTS.
- There is the option to disable the materialisation on creation with specifying WITH [ NO ] DATA.
- If there is no data in the materialised view, the view cannot be queried.
- If there is a materialization in progress the view returns the latest completed materialization.
- There is information in an information schema table that hold the last timestamp the view has been materialized. 

```sql
CREATE MATERIALIZED VIEW [ IF NOT EXISTS ] table_name
    [ (column_name [, ...] ) ]
    [ WITH ( storage_parameter [= value] [, ... ] ) ]
    AS query
    [ WITH [ NO ] DATA ]
```
- There is no automatic refresh of a materialized view
- If the table or views that the materialized view is referring to don’t exist anymore or don’t hold the requested data (e.g. columns missing), the view does not get invalidated, it just can’t get updated anymore.
- There is no delta-refresh or continuous refresh. Every refresh is a complete new materialization.

![image](https://github.com/crate/crate/assets/23557193/72f53fa5-6eda-4ea6-b974-67822e2e71c9)

**Comment by hlcianfagna** [comment]
Related with 
https://github.com/crate/crate/issues/10731
and
https://github.com/crate/crate/issues/12218

**Comment by robd003** [comment]
Being able to define an automatic refresh interval (once every 15 mins / once every hour) would be really helpful. Being able to tweak the refresh rate w/o having to destroy & recreate the materialized view would also be great.

**Comment by helenap** [comment]
I could **really use** materialized views for complicated computations that are used repeatedly in other queries. Could you please make this a priority in 5.8?

**Comment by hlcianfagna** [comment]
Thank you for the feedback @helenap , while waiting for this to be implemented I would suggest either using the new [SQL job scheduler in CrateDB Cloud](https://cratedb.com/blog/introducing-scheduled-jobs-and-cloud-query-console) or [dbt's table/incremental materializations](https://community.cratedb.com/t/using-dbt-with-cratedb/1566).

**Comment by robd003** [comment]
@mfussenegger any idea what the roadmap is for Materialized Views?

**Comment by mfussenegger** [comment]
> any idea what the roadmap is for Materialized Views?

I don't really have more info on this as we generally only plan ahead one release. That's currently 6.2.0 and materialized views won't be part of it. 6.2.0 was/is more focused on addressing some performance problems seen on clusters running in the crate cloud.

But it's generally a often requested feature and we did a lot of work in the direction of supporting materialized views as part of 6.0 and 6.1, so it's possible that we will continue that work within the 6.3.0 cycle. But also possible that until then other more pressing matters will have popped up.

**Comment by robd003** [comment]
@mfussenegger I've built an implementation of Materialized Views where it just creates a new CrateDB table and populates it using `INSERT ... SELECT` based on the `VIEW` query.

When `REFRESH MATERIALIZED VIEW {name}` is called then a new table is created with the updated data and then does an atomic swap with the original view table and then deletes the original view table.

I'm going to clean this up over the weekend, but I can open a pull request if you guys are interested in adding this to the 6.5 release.

One thing that would be really useful would to add a `cron` type system to CrateDB so that Materialized Views could be automatically updated.

**Comment by mfussenegger** [comment]
> where it just creates a new CrateDB table and populates it using INSERT ... SELECT based on the VIEW query.

I'm not sure creating a regular table is the way to go, as we'll likely want to have a different set of supported operations (e.g. no direct insers from user).
I think to implement this we'll want to have some kind of `RelationMetadata.MaterializedView` that's linked to an index.

The refresh materialized view can then create a new index, and once populated update the `indexUUIDs` of the MaterializedView metadata and drop the old index/partition.

**Comment by robd003** [comment]
@mfussenegger Here's what I've been working on https://github.com/crate/crate/pull/19962
